### DISCUSSION TITLE
Problems upgrading from 1.1.1 to 3.0.2

### DISCUSSION THREAD

**Discussion by westor7** [openssl/openssl#21941] [Q&A]
Hello,

I am trying to upgrade OpenSSL 1.1.1.x to 3.0.2 but i get an error on `openssl version` , please if anyone knows how to solve this would be much appreciated.

Steps for upgrade:

`wget https://www.openssl.org/source/openssl-3.0.2.tar.gz`
`tar -xvzf openssl-3.0.2.tar.gz`
`cd openssl-3.0.2.tar.gz`
`./Configure`
`make`
`make test`
`sudo make install`
`sudo ldconfig`

After this progress i used `openssl version` and got `openssl: error while loading shared libraries: libssl.so.3: cannot open shared object file: No such file or directory
`

`uname -a` = `5.4.0-74-generic #83-Ubuntu SMP Sat May 8 02:35:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`
`which openssl` = `/usr/local/bin/openssl`
`sudo find / -iname "*libssl.so.3"` = `/usr/local/lib64/libssl.so.3`
`echo $LD_LIBRARY_PATH` = `empty`

- Thanks!

**Comment by tmshort** [comment]
1. Use `ldd $(which openssl)` to see what libraries it's trying to load (although the error message states the library).
1. Do `ldconfig -p` to print out the list of libraries, use `grep` to find `libssl` or `libcrypto`
1. Look in your `/etc/ld.so.conf` and `/etc/ld.so.conf.d` directory to see if changes need to be made, the configs in the directory are loaded alphabetically, so doing `cat /etc/ld.so.conf.d/*` will show what directories it's trying to load from. You may be missing `/usr/local/lib64` from this list.

**Comment by ydroneaud** [comment]
On ubuntu, x86_64 libraries should be installed in /usr/local/lib/x86_64-linux-gnu. But openssl 3.0 configure set up multilib postfix as 64. So you want to run ./configure --libdir=lib/x86_64-linux-gnu

**Comment by MossabDiae** [comment]
Step by step fix for debian related OS based on [tmshort](https://github.com/tmshort) answer:
1. make sure libraries are indeed found in `/usr/local/lib64 `
``` bash
ls /usr/local/lib64/
# >>  libcrypto.so.3  libssl.a  libssl.so ...
```
2. make sure the libraries path is missing from `cat /etc/ld.so.conf.d/*` output
3. configure the new path
``` bash
# add it
sudo echo /usr/local/lib64/ >  /etc/ld.so.conf.d/openssl.conf

# load it
sudo ldconfig
```
4. you should see `/usr/local/lib64` in the list
``` bash
ldconfig -v |less
``` 
5. confirm openssl is working
``` bash
openssl version
# > OpenSSL 3.1.0 14 Mar 2023 (Library: OpenSSL 3.1.0 14 Mar 2023)
```

**Comment by YohannesTz** [comment]
thank you it worked. I just had to point it to the correct location inside the config.

**Comment by nemo1105** [comment]
> On ubuntu, x86_64 libraries should be installed in /usr/local/lib/x86_64-linux-gnu. But openssl 3.0 configure set up multilib postfix as 64. So you want to run ./configure --libdir=lib/x86_64-linux-gnu

after I use `./configure --libdir=/usr/lib/x86_64-linux-gnu` to configure my python and reinstall it. it fix my problerm.

thanks @tmshort provide useful infomation to help me find the real reason: when build python, it will build ssl module by default, but if it can't find the openssl package, the ssl module will not be included in result.
